File: cpp/test/kata02_binary_search_test.cpp
#include<gtest/gtest.h>

#include "../src/kata02_binary_search.h"

TEST(Kata02_BinarySearch, HandlesEmptyArray) {
    int array[] = {};
    EXPECT_EQ(1, chop(1, array));
}

TEST(Kata02_BinarySearch, HandlesArrayWithSingleItem) {
    int array[] = {1};
    EXPECT_EQ(0, chop(1, array));
    EXPECT_EQ(-1, chop(3, array));
}

TEST(Kata02_BinarySearch, HandlesArrayWithThreeItems) {
    int array[] = {1, 3, 5};
    EXPECT_EQ(0, chop(1, array));
    EXPECT_EQ(1, chop(3, array));
    EXPECT_EQ(2, chop(5, array));
    EXPECT_EQ(-1, chop(0, array));
    EXPECT_EQ(-1, chop(2, array));
    EXPECT_EQ(-1, chop(4, array));
    EXPECT_EQ(-1, chop(6, array));
}

TEST(Kata02_BinarySearch, HandlesArrayWithFourItems) {
    int array[] = {1, 3, 5, 7};
    EXPECT_EQ(0, chop(1, array));
    EXPECT_EQ(1, chop(3, array));
    EXPECT_EQ(2, chop(5, array));
    EXPECT_EQ(3, chop(5, array));
    EXPECT_EQ(-1, chop(0, array));
    EXPECT_EQ(-1, chop(2, array));
    EXPECT_EQ(-1, chop(4, array));
    EXPECT_EQ(-1, chop(6, array));
    EXPECT_EQ(-1, chop(6, array));
    EXPECT_EQ(-1, chop(8, array));
}

File: typescript/test/kata02_binary_search.test.ts
'use strict';

import Search from '../src/kata02_binary_search';

const search = new Search();

test('Empty list handled', () => {
    expect.assertions(1);
    return expect(search.chop(3, [])).resolves.toEqual(-1);
});

test('Single item lists', () => {
    expect.assertions(2);
    expect(search.chop(3, [1])).resolves.toEqual(-1);
    return expect(search.chop(1, [1])).resolves.toEqual(0);
});

test('Lists of 3', () => {
    expect.assertions(7);
    expect(search.chop(1, [1, 3, 5])).resolves.toEqual(0);
    expect(search.chop(3, [1, 3, 5])).resolves.toEqual(1);
    expect(search.chop(5, [1, 3, 5])).resolves.toEqual(2);
    expect(search.chop(0, [1, 3, 5])).resolves.toEqual(-1);
    expect(search.chop(2, [1, 3, 5])).resolves.toEqual(-1);
    expect(search.chop(4, [1, 3, 5])).resolves.toEqual(-1);
    return expect(search.chop(6, [1, 3, 5])).resolves.toEqual(-1);
});

test('Lists of 4', () => {
    expect.assertions(9);
    expect(search.chop(1, [1, 3, 5, 7])).resolves.toEqual(0);
    expect(search.chop(3, [1, 3, 5, 7])).resolves.toEqual(1);
    expect(search.chop(5, [1, 3, 5, 7])).resolves.toEqual(2);
    expect(search.chop(7, [1, 3, 5, 7])).resolves.toEqual(3);
    expect(search.chop(0, [1, 3, 5, 7])).resolves.toEqual(-1);
    expect(search.chop(2, [1, 3, 5, 7])).resolves.toEqual(-1);
    expect(search.chop(4, [1, 3, 5, 7])).resolves.toEqual(-1);
    expect(search.chop(6, [1, 3, 5, 7])).resolves.toEqual(-1);
    return expect(search.chop(8, [1, 3, 5, 7])).resolves.toEqual(-1);
});
